airflow/contrib/operators/sagemaker_create_training_job_operator.py

```diff
@@ -0,0 +1,98 @@
+# -*- coding: utf-8 -*-
+#
+# Licensed to the Apache Software Foundation (ASF) under one
+# or more contributor license agreements.  See the NOTICE file
+# distributed with this work for additional information
+# regarding copyright ownership.  The ASF licenses this file
+# to you under the Apache License, Version 2.0 (the
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+
+from airflow.contrib.hooks.sagemaker_hook import SageMakerHook
+from airflow.models import BaseOperator
+from airflow.utils import apply_defaults
+from airflow.exceptions import AirflowException
+
+
+class SageMakerCreateTrainingJobOperator(BaseOperator):
+
+    """
+       Initiate a SageMaker training
+
+       This operator returns The ARN of the model created in Amazon SageMaker
+
+       :param training_job_config:
+       The configuration necessary to start a training job (templated)
+       :type training_job_config: dict
+       :param region_name: The AWS region_name
+       :type region_name: string
+       :param sagemaker_conn_id: The SageMaker connection ID to use.
+       :type aws_conn_id: string
```

Comment:
Should be `sagemaker_conn_id`

Reply:
Hi Fokko, 
Thank you so much for your review. I really appreciate your feedback. I didn't figure out how to reply to your request, so I'll just reply to you here. The main reason why I separate it to operator and sensor is that the success of the training job have two stages: successfully kick off a training job, and the training job successfully finishes. The operator tells about the first status, and the sensor tells the latter one. Also, since a training job is hosted at an AWS instance, not the instance that is hosting Airflow, so this way, other operators can set upstream to the operator, rather than the sensor, if they aren't dependent on the model actually being created. Also, by using the sensor, users can set parameters like poke_interval, which makes more sense for a sensor rather than an operator.

Reply:
Hi Keliang, thanks for explaining the Sagemaker process. I think it is very similar to for example the Druid hook that we have: https://github.com/apache/incubator-airflow/blob/master/airflow/hooks/druid_hook.py#L93

This hook will kick of a job using a HTTP POST of a json document to the druid cluster, and make sure that it receives a http 200. And then it will continue to poll the job by invoking the API periodically.

Reply:
@Fokko - To further add to Keliang explanation of separating the operator to kick off the job and sensor to monitor the job is to provide flexibility to users. Based on their use case(s), they may have the requirements to kick of multiple jobs/tasks in parallel and then monitor the completion of all Amazon Sagemaker job(s) downstream. Some these jobs may take hours and we don't want to hold the pipeline to initiate other downstream jobs hampering the users from meeting their required SLA. Since there are many other known and unknown scenarios, we are careful not to club both the initialization and monitoring of the job. 

The design is similar to the Amazon EMR - https://github.com/apache/incubator-airflow/blob/master/airflow/contrib/example_dags/example_emr_job_flow_manual_steps.py

Reply:
Hi @Fokko 

I totally agree with what @srrajeev-aws mentioned here. And as a developer and maintainer of SageMaker sdk, I can confirm that a lot users will use SageMaker on data with large size which causes

1. Training is time-consuming.
2. Users may run parallel training jobs
3. Users may want to manage ongoing training jobs. 

So an asynchronous way of training would be preferred than a blocking one.

Reply:
@srrajeev-aws In this case you would just kick off multiple operators in parallel. This is inherent of the concept of a DAG, if the training jobs don't have any dependencies on each other, they will just run in parallel. The only flexibility that the decoupling of the kicking of the job, and monitoring the job is in the case when you don't care about the outcome of the job. This is also analoge to Druid, an indexing job can take up to a couple of hours.
Having a separate operator and sensor would make the DAGs unnecessarily complicated, since in practice you will always use them as a pair.

Reply:
@Fokko 
Thank you very much for your reply. I agree with you that multiple jobs can still run in parallel, but to my understanding, there's also an option to run in sequential.  More importantly, I think one thing that is very essential is that, what if there's already a training job running (whether initiated from Airflow or other means) before a dag is scheduled to run? Since a training job can takes days to finish, I think it might be a potential use case for Airflow users. As a result, a DAG can start with a sensor, and have other nodes dependent on the sensor. It would not be possible to accomplish this if operator, and sensor are coupled together.

Reply:
@troychen728 The use case you mention justifies adding the sensor, but not removing the functionality from the operator.

Running sequentially is not really suited for production workflows, it should not be a concern.

Having a sensor per operator problematic though (for normal usage), hence keeping the functionality in the operator sounds the most sensible to me.

Reply:
@Fokko @gglanzani 
I agree with you that giving user an easier way is always helpful, so I updated the code, and added an wait option in operator. The default is True, and if it is not set to False by user, the operator will only signal success after the training job has finished. The control logic should be very similar to that of Druid hook you mentioned before.